### ISSUE TITLE
🪡 IGLA Experiment Matrix — Deep Benchmark Protocol (IGLA-JEPA-T, IGLA-Muon, IGLA-GF16)

### ISSUE THREAD

**Issue by gHashTag** [gHashTag/trios#119]
# 🪡 IGLA — Игла Кощея Бессмертного

**ИГЛА** (рус. "игла") = Needle in the Haystack. Если найдём — Parameter Golf наш.
Научный код-нейм серии экспериментов Trinity для OpenAI Parameter Golf (deadline Apr 30, 2026).

## 🎯 Миссия

Найти комбинацию техник, которая даёт **BPB ≤ 1.10** при ограничениях:
- ≤ 16 MB (код + веса, zstd-22 compressed)
- ≤ 10 минут на 8×H100
- Byte-level, FineWeb-EDU validation

---

## 🧬 Три ветви экспериментов

| Ветвь | Суть | Научный префикс |
|---|---|---|
| **IGLA-JEPA-T** | Joint-Embedding Predictive (LeCun 2023) + T-trinity 3^k dims | Self-supervised latent prediction |
| **IGLA-Muon** | Newton-Schulz ortho optimizer (Jordan 2024) + φ-schedule | Orthogonalized gradient flow |
| **IGLA-GF16** | Golden-ratio base-φ float quantization + zstd-22 | Log-normal weight distribution |

---

### Легенда колонок

| Col | Meaning |
|---|---|
| **ID** | IGLA-{branch}-{NNN} |
| **Arch** | layers × d_model × heads |
| **Opt** | AdamW / Muon / Muon+φ |
| **Quant** | BF16 / FP8 / INT6 / GF16 / BitNet1.58 |
| **Attn** | Vanilla / Flex / LSWA / Sparse-Gate |
| **Pos** | RoPE / NoPE / φ-RoPE |
| **BPB** | bits-per-byte на FineWeb-val (lower = better) |
| **Size** | zstd-22 bytes total |
| **Wall** | minutes on 8×H100 |
| **Recall** | NIAH @ 4K ctx (%) |
| **∇-norm** | median grad norm (stability) |
| **Seed σ** | std across 3 seeds |

---

### Phase A — Baseline Calibration (ID 001–005)

| ID | Arch | Opt | Quant | Attn | Pos | Target BPB | Purpose |
|---|---|---|---|---|---|---|---|
| IGLA-BASE-001 | 6×512×8 | AdamW | BF16 | Vanilla | RoPE | 1.2244 | Reproduce OpenAI baseline |
| IGLA-BASE-002 | 6×512×8 | AdamW | BF16 | Vanilla | RoPE | 1.2244 | Confirm 3-seed variance < 0.005 |
| IGLA-BASE-003 | 11×512×8 | AdamW | INT6 | SWA | RoPE | ~1.15 | Reproduce thwu1 stack |
| IGLA-BASE-004 | 9×729×9 | AdamW | BF16 | Vanilla | RoPE | ? | First **3^k Trinity** arch |
| IGLA-BASE-005 | 9×729×9 | AdamW | BF16 | Vanilla | RoPE | ? | Zero-init projections (anti-divergence) |

### Phase B — IGLA-Muon Series (ID 101–110)

| ID | Arch | Opt | Quant | Attn | Extra | Expected gain |
|---|---|---|---|---|---|---|
| IGLA-MUON-101 | 9×729×9 | Muon | BF16 | Vanilla | NS-5 iter, default | +35% speed |
| IGLA-MUON-102 | 9×729×9 | Muon | BF16 | Vanilla | Momentum warmup 0→0.95 | stability |
| IGLA-MUON-103 | 9×729×9 | Muon+φ | BF16 | Vanilla | LR = base·φ^(-step/total) | convergence |
| IGLA-MUON-104 | 9×729×9 | Muon | BF16 | Vanilla | QK-Norm + RMSNorm + ReLU² | +10% BPB |
| IGLA-MUON-105 | 9×729×9 | Muon | BF16 | Vanilla | Logit softcap 30·tanh(x/30) | no-NaN |
| IGLA-MUON-106 | 9×729×9 | Muon | FP8-head | Vanilla | E4M3 fwd / E5M2 bwd | +8% speed |
| IGLA-MUON-107 | 9×729×9 | Muon | FP8-head | FlexAttn | Long-short 1024/2048 window | +15% speed |
| IGLA-MUON-108 | 9×729×9 | Muon | FP8-head | FlexAttn | + Value Embeddings (first/last 3) | +3% BPB |
| IGLA-MUON-109 | 9×729×9 | Muon | FP8-head | FlexAttn | + U-Net skips + X0-mixin | +5% BPB |
| IGLA-MUON-110 | 9×729×9 | Muon | FP8-head | Sparse-Gate | + BOS-sink removal | +2% BPB |

### Phase C — IGLA-JEPA-T Series (ID 201–208)

| ID | Arch | Pretrain | Quant | Mask ratio | Latent dim | Purpose |
|---|---|---|---|---|---|---|
| IGLA-JEPA-201 | 9×729×9 | T-JEPA | BF16 | 0.15 | 243 | Baseline JEPA |
| IGLA-JEPA-202 | 9×729×9 | T-JEPA | BF16 | 0.40 | 243 | High mask (MAE-style) |
| IGLA-JEPA-203 | 9×729×9 | T-JEPA+NCA | BF16 | 0.25 | 243 | NCA warmup → JEPA main |
| IGLA-JEPA-204 | 9×729×9 | T-JEPA | BF16 | 0.25 | 729 | Fibonacci latent dim |
| IGLA-JEPA-205 | 9×729×9 | T-JEPA+Muon | BF16 | 0.25 | 243 | **JEPA × Muon synergy** |
| IGLA-JEPA-206 | 9×729×9 | T-JEPA | BF16 | φ-schedule | 243 | Mask ratio decay by 1/φ |
| IGLA-JEPA-207 | 9×729×9 | T-JEPA-EMA | BF16 | 0.25 | 243 | EMA teacher 0.997 |
| IGLA-JEPA-208 | 9×729×9 | I-JEPA→T-JEPA | BF16 | 0.25 | 243 | Curriculum transfer |

### Phase D — IGLA-GF16 Series (ID 301–308)

| ID | Arch | Opt | Quant | Compress | Params | Target size |
|---|---|---|---|---|---|---|
| IGLA-GF16-301 | 9×729×9 | Muon | GF16 | zstd-22 | 33M | 15.8 MB |
| IGLA-GF16-302 | 9×729×9 | Muon | GF16 QAT | zstd-22 | 33M | STE gradients |
| IGLA-GF16-303 | 9×729×9 | Muon | GF16 + GPTQ-lite | zstd-22 | 33M | clip search |
| IGLA-GF16-304 | 11×729×9 | Muon | BitNet 1.58 | zstd-22 | **67M** | ternary {-1,0,1} |
| IGLA-GF16-305 | 9×729×9 | Muon | GF16 mixed (heads BF16) | zstd-22 | 33M | hybrid |
| IGLA-GF16-306 | 9×729×9 | Muon | GF16 + Vocab pad 768 | zstd-22 | 33M | tensor-core friendly |
| IGLA-GF16-307 | 9×729×9 | Muon | GF16 + 2D-block FP8 GEMM | zstd-22 | 33M | Triton kernel |
| IGLA-GF16-308 | 11×729×9 | Muon | BitNet 1.58 + TWEO | zstd-22 | 67M | outlier suppression |

### Phase E — Ensemble & Final (ID 401–405)

| ID | Strategy | Components | Expected BPB |
|---|---|---|---|
| IGLA-FINAL-401 | Best single (Muon+FP8+Flex+GF16) | 110 + 301 | 1.11 |
| IGLA-FINAL-402 | JEPA→Standard finetune | 205 → 110 | 1.10 |
| IGLA-FINAL-403 | BitNet deep | 308 | 1.10 |
| IGLA-FINAL-404 | Ensemble logit avg | 401+402+403 | 1.09 |
| IGLA-FINAL-405 | **SUBMIT** | best of 404, 3 seeds | **≤ 1.10** |

---

### Primary (gating)
- **BPB FineWeb-val** — shard 000, stride-64 sliding window
- **Compressed size** (zstd-22, code+weights)
- **Wall-clock** on 8×H100 SXM (or CPU-equiv × scaling factor)

### Secondary (diagnostic)
- **NIAH recall** @ contexts {1K, 2K, 4K, 8K} — "игла в стоге"
- **Gradient norm median / p99** — stability
- **Activation max** — outlier detection (TWEO threshold < 20)
- **Loss curve smoothness** — TV(loss) / loss_final
- **Throughput** — tokens/sec
- **Memory peak** — MiB
- **Seed variance σ** — across seeds {42, 1337, 27}

### Tertiary (forensic)
- **Per-layer BPB contribution** — which layer does the work
- **Attention entropy** — head utilization
- **Weight histogram** — log-normal fit quality (для GF16 justification)
- **Gradient spectrum** — Muon Newton-Schulz convergence
- **BOS-sink ratio** — % attention on token 0

---

## 📐 Protocol — единый для всех

```
1. Fixed seed {42, 1337, 27} — always 3 runs
2. Data: FineWeb-EDU shard 000, byte-level, 10B tokens max
3. Batch: 524288 tokens global (fixed across experiments)
4. Steps: linear schedule, 6000 steps = 10 min budget
5. Eval every 500 steps on held-out 10M bytes
6. Log: BPB, grad_norm, act_max, throughput → wandb + local JSONL
7. Gate: if grad_norm > 10 for 10 steps → kill + log failure
8. Artifact: save best checkpoint + submission.tar.zst
9. Report: ring-NNN-report.md with plots (loss, BPB, grad, act)
```

---

## 🎯 Acceptance Criteria для IGLA series

| Gate | Threshold | Action if fail |
|---|---|---|
| G1: Reproduces baseline | BPB 1.2244 ± 0.005 | Debug data pipeline |
| G2: Muon works | BPB < 1.20, no NaN | Fallback AdamW |
| G3: Trinity 3^k stable | no divergence 9 layers | Zero-init proj fix |
| G4: GF16 roundtrip | weight error < 1% | Adjust mantissa |
| G5: Size budget | zstd-22 ≤ 16 MB | Cut layers or d_model |
| G6: Time budget | ≤ 10 min on 8×H100 | Cut steps or FP8 |
| G7: SOTA threshold | BPB ≤ 1.15 | Stack more tier-S tricks |
| G8: Submit ready | 3 seeds σ < 0.005 | Re-run with different seeds |

---

## 🗓️ Timeline (11 days to Apr 30)

| Day | Phase | Experiments |
|---|---|---|
| Apr 20 | A | 001–005 baseline |
| Apr 21 | B | 101–105 Muon core |
| Apr 22 | B | 106–110 Muon + FP8/Flex |
| Apr 23 | C | 201–204 JEPA baseline |
| Apr 24 | C | 205–208 JEPA advanced |
| Apr 25 | D | 301–304 GF16 / BitNet |
| Apr 26 | D | 305–308 GF16 optimized |
| Apr 27 | E | 401–403 finalists |
| Apr 28 | E | 404 ensemble |
| Apr 29 | E | 405 submit prep, 3 seeds |
| Apr 30 | 🚀 | **SUBMIT** |

---

## 📚 Scientific References (10 papers)

1. **Muon Optimizer** — Jordan et al. 2024, NanoGPT speedrun
2. **T-JEPA** — LeCun et al. 2023, "A Path Towards Autonomous ML"
3. **BitNet b1.58** — Ma et al. 2024, "The Era of 1-bit LLMs"
4. **TWEO** — 2025, "Transformers Without Extreme Outliers" (FP8 pretrain)
5. **FlexAttention** — PyTorch 2024, long-short sliding window
6. **Value Embeddings** — Zhou et al. 2024
7. **NIAH Benchmark** — Kamradt 2023, gkamradt/LLMTest_NeedleInAHaystack
8. **InfiniRetri** — 2025, 100% recall @ 1M tokens on 0.5B model
9. **Gemma 2 softcap** — Google 2024, logit stabilization
10. **Modded-NanoGPT** — KellerJordan 2024-2025, speedrun record

---

## 🪡 Why IGLA = бессмертие

> В русских сказках Кощей Бессмертный прячет смерть в **игле**, игла в яйце, яйцо в утке, утка в зайце, заяц в сундуке, сундук на дубе. 6 уровней вложенности.

Наши 6 уровней:
1. 🌳 **FineWeb корпус** (дуб)
2. 📦 **16 MB лимит** (сундук)  
3. 🐰 **9-layer trinity 3^k** (заяц)
4. 🦆 **Muon + JEPA latents** (утка)
5. 🥚 **GF16 weight encoding** (яйцо)
6. 🪡 **BPB ≤ 1.10** (ИГЛА)

**Найдёшь иглу — конкурс твой, хозяин вечен.** 🏆

---

/label ~igla ~parameter-golf ~experiments ~benchmark
/cc agents: A (Parameter Golf), C (Muon R12), D (GF16), E (JEPA)

**Comment by gHashTag** [comment]
## 🎯 Night-shift dispatch → @trinity-researcher

**Priority:** P1 (High, связан с #110 deadline Apr 30)
**Bridge:** BrowserOS :9002
**Constitutional:** L21–L25

### Task
Фиксирующий **benchmark protocol spec** для трёх ветвей IGLA (JEPA-T, Muon, GF16), с воспроизводимыми критериями для #110 submission.

### Deliverables
1. `docs/igla/benchmark-protocol.md` — фиксирует:
   - Трен-вал split (FineWeb-EDU, seeds, хэши).
   - Hardware matrix (1×H100 smoke, 8×H100 full).
   - Metric set: BPB, tokens/s, param count, compressed size (zstd-22), wall-clock.
   - Stopping criteria: 10 min wall-clock или val BPB plateau < 0.001 по последним 3 evals.
   - Repro hash: git SHA + config SHA + data SHA в каждом run manifest.
2. `configs/igla/{jepa-t,muon,gf16}.toml` — базовые конфиги для трёх ветвей.
3. `scripts/igla/verify-repro.sh` — проверяет, что run воспроизводим побитово на том же seed.
4. Координация с #110: этот protocol становится source of truth для `configs/param-golf/run-1.toml`.
5. PR → branch `igla/protocol-v1` → `main`.

### Constraints
- Append-only (L21): переписывать protocol нельзя, только v2/v3 новыми файлами.
- Все числа — с 95% CI по 3+ seeds.
- Никаких cherry-picked metrics, всё runs в manifest.

### Acceptance
- [ ] Protocol md коммит
- [ ] 3 config файла
- [ ] verify-repro.sh проходит на mini fixture
- [ ] Linked в #110
- [ ] PR open, CI green

---
*Dispatched by Comet (night-shift). Этот protocol — камень в фундамент Parameter Golf submission.*

**Comment by gHashTag** [comment]
Requires GPU training infrastructure for deep benchmarks. Parameter Golf ended. Closing.
